### PR TITLE
Add notifications for findings-data-import Lambda processing failures

### PR DESCRIPTION
**Pull request by mcdonnnj**
#  #

## 🗣 Description ##

This pull request adds all of the Terraform necessary to send notifications if processing a findings JSON with [cisagov/findings-data-import-lambda] fails.

## 💭 Motivation and context ##

This addition will add the ability to notify operations staff if there is a failure when processing a findings JSON. This will help alert the appropriate team if there is an issue with the data or the Lambda.

## 🧪 Testing ##

Automated tests pass. I applied these changes in my development environment and was able to confirm I received an email notification if I uploaded an invalid JSON to the bucket in my environment and the Lambda failed when processing.

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] All new and existing tests pass.

[cisagov/findings-data-import-lambda]: https://github.com/cisagov/findings-data-import-lambda

### CODE REVIEW

File: terraform/fdi_lambda.tf [hunk 1]
```diff
@@ -196,16 +196,27 @@ resource "aws_lambda_permission" "fdi_lambda_allow_bucket" {
   source_arn    = data.aws_s3_bucket.findings_data.arn
 }
 
-# Create the notification that triggers our Lambda function to run whenever
-# an object is created in our findings data bucket
+# Create the notification configuration for the findings data bucket
 resource "aws_s3_bucket_notification" "fdi_lambda" {
   bucket = data.aws_s3_bucket.findings_data.id
 
+  # Trigger the appropriate lambda function whenever an object with the configured
```
Comment (dav3r) [inline terraform/fdi_lambda.tf]: Keep the capitalization train rolling... 🚂 
```suggestion
  # Trigger the appropriate Lambda function whenever an object with the configured
```

**Review by dav3r — APPROVED** [review]
Looks great to me! ⭐

File: terraform/fdi_failure_sns_topic.tf [hunk 2]
```diff
@@ -0,0 +1,47 @@
+# ------------------------------------------------------------------------------
+# Create the SNS topic that allows email to be sent when the findings-data-import
+# Lambda has a failure when processing an uploaded JSON.
+# ------------------------------------------------------------------------------
+
+resource "aws_sns_topic" "fdi_failure_alarm" {
+  name         = format("fdi-failure-alarms_%s", local.production_workspace ? "production" : terraform.workspace)
+  display_name = "findings-data-import_failure_alarm"
```
Comment (jsf9k) [inline terraform/fdi_failure_sns_topic.tf:8]: Should we alphabetize these?

File: terraform/fdi_failure_sns_topic.tf [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  statement {
+    effect = "Allow"
+
+    actions = [
+      "sns:Publish",
+    ]
+
+    principals {
+      type        = "Service"
+      identifiers = ["s3.amazonaws.com"]
```
Comment (jsf9k) [inline terraform/fdi_failure_sns_topic.tf:23]: Should we alphabetize these?

File: terraform/fdi_failure_sns_topic.tf [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    resources = [aws_sns_topic.fdi_failure_alarm.arn]
+
+    condition {
+      test     = "ArnEquals"
+      variable = "aws:SourceArn"
+      values   = [data.aws_s3_bucket.findings_data.arn]
```
Comment (jsf9k) [inline terraform/fdi_failure_sns_topic.tf:31]: Should we alphabetize these?

File: terraform/fdi_failure_sns_topic.tf [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+# Allow the findings data bucket (and only the findings data bucket) to publish
+# notifications to the fdi failure alarm SNS topic.
+data "aws_iam_policy_document" "fdi_failure_alarm" {
+  statement {
+    effect = "Allow"
+
+    actions = [
+      "sns:Publish",
+    ]
```
Comment (jsf9k) [inline terraform/fdi_failure_sns_topic.tf:19]: Should we alphabetize these?

File: terraform/fdi_lambda.tf [hunk 6]
```diff
@@ -196,16 +196,27 @@ resource "aws_lambda_permission" "fdi_lambda_allow_bucket" {
...
+  # suffix is created in the bucket.
   lambda_function {
     lambda_function_arn = aws_lambda_function.fdi_lambda.arn
     events              = ["s3:ObjectCreated:Put"]
     filter_suffix       = var.findings_data_input_suffix
```
Comment (jsf9k) [inline terraform/fdi_lambda.tf:208]: Should we alphabetize these?

File: terraform/fdi_lambda.tf [hunk 7]
```diff
@@ -196,16 +196,27 @@ resource "aws_lambda_permission" "fdi_lambda_allow_bucket" {
...
   }
+
+  # Notify the appropriate SNS topic for fdi failures whenever an object is created
+  # with the configured prefix and suffix.
+  topic {
+    topic_arn     = aws_sns_topic.fdi_failure_alarm.arn
+    events        = ["s3:ObjectCreated:Copy"]
+    filter_prefix = var.findings_data_import_lambda_failure_prefix
+    filter_suffix = var.findings_data_import_lambda_failure_suffix
```
Comment (jsf9k) [inline terraform/fdi_lambda.tf:217]: Should we alphabetize these?

**Review by jsf9k — APPROVED** [review]
Great addition!  I have no comments other than a few suggestions about alphabetization.

File: terraform/fdi_failure_sns_topic.tf (hunk 2 above)
Comment (mcdonnnj) [reply terraform/fdi_failure_sns_topic.tf:8]: If you are amenable I would like to punt that off into a separate PR to reformat _all_ of the Terraform as one cohesive piece of work.

Comment (jsf9k) [reply terraform/fdi_failure_sns_topic.tf:8]: Sure

File: terraform/fdi_failure_sns_topic.tf (hunk 3 above)
Comment (jsf9k) [reply terraform/fdi_failure_sns_topic.tf:23]: See [here](https://github.com/cisagov/cyhy_amis/pull/646#discussion_r1161825858).

File: terraform/fdi_failure_sns_topic.tf (hunk 5 above)
Comment (jsf9k) [reply terraform/fdi_failure_sns_topic.tf:19]: See [here](https://github.com/cisagov/cyhy_amis/pull/646#discussion_r1161825858).

File: terraform/fdi_failure_sns_topic.tf (hunk 4 above)
Comment (jsf9k) [reply terraform/fdi_failure_sns_topic.tf:31]: See [here](https://github.com/cisagov/cyhy_amis/pull/646#discussion_r1161825858).

File: terraform/fdi_lambda.tf (hunk 6 above)
Comment (jsf9k) [reply terraform/fdi_lambda.tf:208]: See [here](https://github.com/cisagov/cyhy_amis/pull/646#discussion_r1161825858).

File: terraform/fdi_lambda.tf (hunk 7 above)
Comment (jsf9k) [reply terraform/fdi_lambda.tf:217]: See [here](https://github.com/cisagov/cyhy_amis/pull/646#discussion_r1161825858).

**Pull request merged**